### ISSUE TITLE
DOC502 (and by extension DOC503) should not be seen if called functions are already annotated with the given Exception

### ISSUE THREAD

**Issue by AndreiMiculita** [jsh9/pydoclint#238]
DOC502:	Function/method has a “Raises” section in the docstring, but there are not “raise” statements in the body
DOC503:	Exceptions in the “Raises” section in the docstring do not match those in the function body

Suppose we have:

```python
def foo(x):
    """
    Raises:
        ValueError: when bound check failed
    """
    checkAgainstBounds(x)
    manipulate(x)

def checkAgainstBounds(x):
    """
    Raises:
        ValueError: when X is not within bounds
    """
    if x < 0 or x > 100:
        raise ValueError(f"value {x} is out of bounds")
    maybeSomeOtherFunction()
```

This triggers `DOC502`.
But I propose this should be valid as the ValueError is explicitly declared in the docstring of `checkAgainstBounds`. It's not raised in `foo`, so to fix the DOC502 we need to explicitly re-raise it:

```python
def foo(x):
    """
    Raises:
        ValueError: when bound check failed
    """
    try:
        checkAgainstBounds(x)
    except(e):
        raise ValueError(e.message) from e  # or something like that?
    manipulate(x)
````

To avoid having to do this, the behavior of DOC502 IMO should be changed so that exceptions can either be explicitly raised OR declared in the `Raises` of a function called in the given function.

It should be enough to only check one level down, i.e. not recurse further down the stack. Because we can assume checkAgainstBounds will itself be checked by the linter (so in this case if `maybeSomeOtherFunction` could raise an exception as well, it would be noticed when checking checkAgainstBounds).

In a large codebase, this would have a cascading effect, basically forcing developers to always either catch an exception in the function or declare it in the docstring. I've found code much easier to maintain and more secure that way when working on PHP (iirc it was this PHPStan rule: https://phpstan.org/blog/bring-your-exceptions-under-control#report-extra-exceptions-in-%40throws-that-aren%E2%80%99t-actually-thrown)

**Comment by AndreiMiculita** [comment]
Follow-up: I notice now it would require changes to DOC501 as well:

DOC501: Function/method has raise statements, but the docstring does not have a “Raises” section

It should basically be: The function/method has raise statements or contains functions annotated with Raises, but they are not caught or declared in the "Raises" section.

**Comment by anicetea** [comment]
The underlying problem here is real: exceptions that escape a function but are not raised in its own body. But I would push back on the approach, on two points.

Covering an exception because any called function documents it casts too wide a net for public callees. If `checkAgainstBounds()` is public and documents `ValueError` in its own Raises section, then having `foo()` document it again is duplication. That contract already lives at the `checkAgainstBounds` boundary. A public method should declare the exceptions that originate in its own implementation, not repeat the documented exceptions of every public method it happens to call. Otherwise the same `ValueError` gets copied into every caller up the chain.

Reading the callee's `Raises` section does not help the case that needs it most. The functions whose escaping exceptions you genuinely cannot see are private helpers, those are often documented lightly or not at all, so there is no `Raises` section to read. Coverage has to come from reading the raise statements inside the private helper, not from its docstring.

So I think the right scoping is private callees only, traced from their raise statements. Public callees own their own contracts and should not cascade. I have written this up as a discussion, since it is more of a design question: #292. Curious what you think.

**Comment by jsh9** [comment]
Hi @AndreiMiculita , I'm sorry I left this issue unattended for more than a year!  I must have missed this issue.

There's a newer issue with the same request: https://github.com/jsh9/pydoclint/issues/283, and there's a new discussion: https://github.com/jsh9/pydoclint/discussions/292

Please feel free to chime in here (in this issue) or in those 2 other places, if you're still interested in this.

I'll think about this too and I'll try to tackle this soon (in the next few weeks).
